File: tic_tac_php/game.php
<?php

interface IGame {
    public function switchCurrentPlayer();
    public function display();
    public function isWinner(): bool;
    public function isDraw(): bool;
    public function isOver(): bool;
    public function makeMove();
}

class Game implements IGame {
    private Player $player1;
    private Player $player2;
    private ?Player $currentPlayer;
    private ?Player $winner;
    private Board $board;

    public function __construct(Player $p1, Player $p2) {
        $this->player1 = $p1;
        $this->player2 = $p2;
        $this->currentPlayer = $p1;
        $this->winner = null;
        $this->board = new Board();
    }

    public function __get($name) {
        return $this->$name;
    }

    public function switchCurrentPlayer() {
        if ($this->currentPlayer == $this->player1) {
            $this->currentPlayer = $this->player2;
        } elseif ($this->currentPlayer == $this->player2) {
            $this->currentPlayer = $this->player1;
        }
    }

    public function display() {
        $this->board->display();
    }

    public function isWinner(): bool {
        return $this->board->checkWinner();
    }

    public function isDraw(): bool {
        return $this->board->isFull();
    }

    public function isOver(): bool {
        if ($this->isWinner()) {
            $this->winner = $this->currentPlayer;
            $this->board->setRemainingCells();
            clearTerminal();
            $this->display();
            $winnerName = $this->winner->name;
            echo "Game over! $winnerName is the winner!\n";
            return true;
        }

        if ($this->isDraw()) {
            $this->winner = null;
            $this->currentPlayer = null;
            echo "Game over! It is a draw";
            return true;
        }

        return false;
    }

    public function makeMove() {
        while (true) {
            try {
                $input = getUserInput($this->currentPlayer->name);
                [$row, $col] = inputToCoords($input);

                $this->board->setCell($row, $col, $this->currentPlayer->mark);
                break;
            } catch (Exception $e) {
                echo $e->getMessage();
            }
        }
    }
}

?>

File: tic_tac_rb/game.rb
require_relative "board.rb"
require_relative "utils.rb"

class Game
    attr_accessor :player1, :player2, :current_player, :winner, :board

    def initialize(player1, player2)
        @player1 = player1
        @player2 = player2
        @current_player = player1
        @winner = nil
        @board = Board.new
    end

    def switch_current_player
        if @current_player.equal?(@player1)
            @current_player = @player2
        elsif @current_player.equal?(@player2)
            @current_player = @player1
        end
    end

    def display
        @board.display
    end

    def is_winner
        @board.check_winner
    end

    def is_draw
        @board.is_full
    end

    def is_over
        if is_draw
            @winner = nil
            @current_player = nil
            puts "Game over! It is a draw!"
            return true
        end

        if is_winner
            @winner = @current_player
            @board.set_remaining_cells
            clear_terminal
            display
            puts "Game over! #{@current_player.name} is the winner"
            return true
        end

        return false
    end

    def make_move
        while true
            begin
                input = get_user_input(@current_player.name)
                row, col = input_to_coords(input)

                @board.set_cell(row - 1, col - 1, @current_player.mark)
                break
            rescue ArgumentError, IndexError => e
                puts e.message
            end
        end
    end

end